### DISCUSSION TITLE
Synology: How do i determine if wireguard is using kernel module or go

### DISCUSSION THREAD

**Discussion by arpitgupta** [qdm12/gluetun#736] [Q&A] [answered]
I understand Synlogy does not natively have the wireguard kernel module. However i installed this https://github.com/runfalk/synology-wireguard which adds wireguard support to synology. So i was thinking that the container should now use the kernel module? Is there a way for me to confirm that? Is there anything in the logs that has that info?

**Comment by qdm12** [comment, marked as answer]
Ok so I went down a long rabbit hole to find the Wireguard library I was using did not automagically use the Kernelspace implementation if it was available.

HOWEVER, I implemented an 'opportunistic' kernel space usage in cfa3bb3b64c73b0525ed71d75c65125703cac069 (latest image) where the kernelspace implementation is used if it is available; otherwise it fallsback on the (previously existing) userspace Go implementation. My apologies for the false claims, I was mistaken. At least it's fixed now :+1: 

Gluetun will now also log out the implementation it's using (userspace or kernelspace)

**Reply by arpitgupta** [reply]
Thanks. I was curious as I had read a while ago that kernel module would give much better performance vs go and thus wanted to know what my nas was using. Will try out the latest image.

**Reply by qdm12** [reply]
Actually please do some quick bandwidth testing with image `:v3.26.0` (userspace) and latest (kernelspace). Try with a vpn server as close to your gluetun container to reduce bandwidth bottlenecks.

I would bet it makes no difference 😄

**Reply by arpitgupta** [reply]
Ya i am seeing similar results i think. I never did enough of them earlier to have a good baseline :). Curious how with the same image i picked up the new code?

I saw the following line in my logs. 
```
2021/12/14 07:31:20 INFO wireguard: Using available kernelspace implementation
```

**Reply by arpitgupta** [reply]
hmm might be getting better speeds. I am constantly seeing 150+ MBit/s. Earlier i would see some results however 70ish range. However i never really did any baselines. May be if we can explicitly select to use kernel vs userpsace we can do some tests, but even then we will be at the mercy of the ISP.

**Reply by qdm12** [reply]
Yes although I'm in the middle of a big config parsing rework, and since kernel is really meant to be at least as fast as userspace, I would rather prefer not to have it as an option.

Just use image `qmcgaw/gluetun:v3.26.0` which uses userspace, using the same VPN server 😉

**Reply by arpitgupta** [reply]
I am confused now :). So userspace == go? I have been using `qmcgaw/gluetun:v3.26.0` so I should be using the go implementation? Or do i have this wrong lol.

**Reply by qdm12** [reply]
Yes you have it right. Userspace is a Go implementation that exists in both v3.26.0 and latest. The kernel implementation is, well, in the kernel so it's independent from Gluetun. However only the latest gluetun image detects it and uses it, not v3.26.0 which only uses the userspace (go) implementation.

**Reply by arpitgupta** [reply]
Ah duh i was not using a tag in my image so it picked latest :). Here are the results for latest vs 3.26.0

Not sure there is much difference, it could just be my IPS being picky :)

```
## Kernel

# 1
Download: 152.45 Mbit/s
Upload: 24.22 Mbit/s

#2
Download: 54.10 Mbit/s
Upload: 27.89 Mbit/s

#3
Download: 66.95 Mbit/s
Upload: 21.80 Mbit/s


## Go v3.26.0

# 1
Download: 66.34 Mbit/s
Upload: 53.30 Mbit/s

#2
Download: 32.39 Mbit/s
Upload: 18.78 Mbit/s


#3
Download: 26.56 Mbit/s
Upload: 13.48 Mbit/s
```
